### PR TITLE
refac(math): refactor prime fields

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR addresses two key areas:

- Fixes the building of prime fields on GPU.
- Refactors prime fields for better consistency and usability.

Initially, our prime fields were constructed based on EC-based prime fields, leading some interfaces (e.g., operator[]()) to not be suitable for other types of prime fields. After internal discussion, we have decided on standardized interfaces for prime fields.

### CODE REVIEW

File: tachyon/math/finite_fields/small_prime_field_mont.h [hunk 1]
```diff
@@ -17,7 +17,6 @@
 #include "tachyon/base/strings/string_number_conversions.h"
 #include "tachyon/base/strings/string_util.h"
 #include "tachyon/build/build_config.h"
-#include "tachyon/math/base/egcd.h"
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/small_prime_field_mont.h:20]: Can you make this a separate commit?

File: tachyon/math/finite_fields/BUILD.bazel [hunk 2]
```diff
@@ -259,7 +259,6 @@ tachyon_cc_library(
         "//tachyon/base/strings:string_number_conversions",
         "//tachyon/base/strings:string_util",
         "//tachyon/build:build_config",
-        "//tachyon/math/base:egcd",
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/BUILD.bazel:262]: Can you make this a separate commit?

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_config.h.tpl [hunk 3]
```diff
@@ -82,4 +83,4 @@ class TACHYON_EXPORT %{class}Config {
 };
 
 }  // namespace %{namespace}
-// clang-format on
+                                                   // clang-format on
```
Comment (TomTaehoonKim) [inline tachyon/math/finite_fields/generator/prime_field_generator/prime_field_config.h.tpl:86]: Please remove whitespaces here.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/finite_fields/goldilocks/BUILD.bazel [hunk 4]
```diff
@@ -37,6 +37,8 @@ generate_large_fft_prime_fields(
     small_subgroup_adicity = ":" + SMALL_SUBGROUP_ADICITY,
     small_subgroup_base = ":" + SMALL_SUBGROUP_BASE,
     subgroup_generator = ":" + SUBGROUP_GENERATOR,
+    # NOTE(chokobole): Polygon Goldilocks doesn't use montgomery reduction by default.
```
Comment (ashjeong) [inline tachyon/math/finite_fields/goldilocks/BUILD.bazel:39]:
```suggestion
# NOTE(chokobole): Polygon's Goldilocks doesn't use montgomery reduction by default.

File: tachyon/math/finite_fields/goldilocks/goldilocks_prime_field_x86_special.h [hunk 5]
```diff
@@ -49,7 +53,9 @@ class PrimeField<_Config, std::enable_if_t<_Config::kIsGoldilocks>> final
 
   static void Init() { VLOG(1) << Config::kName << " initialized"; }
 
-  value_type value() const { return value_; }
+  // NOTE(chokobole): To be consistent with `PrimeField<F>` defined in
+  // prime_field_fallback.h, it returns the value as a `BigInt<1>`.
+  value_type value() const { return BigInt<1>(value_); }
```
Comment (ashjeong) [inline tachyon/math/finite_fields/goldilocks/goldilocks_prime_field_x86_special.h:54]:
```suggestion
  value_type value() const { return value_type(value_); }

File: tachyon/math/elliptic_curves/msm/algorithms/cuzk/cuzk_unittest.cc [hunk 6]
```diff
@@ -72,13 +73,11 @@ TEST_F(CUZKTest, RunWithRandom) {
...
   auto bases = gpu::GpuMemory<bn254::G1AffinePointGpu>::MallocManaged(size);
   for (size_t i = 0; i < bases.size(); ++i) {
-    bases[i] = bn254::G1AffinePointGpu::FromMontgomery(
-        test_set.bases[i].ToMontgomery());
+    bases[i] = ConvertPoint<bn254::G1AffinePointGpu>(test_set.bases[i]);
   }
   auto scalars = gpu::GpuMemory<bn254::FrGpu>::MallocManaged(size);
   for (size_t i = 0; i < scalars.size(); ++i) {
-    scalars[i] =
-        bn254::FrGpu::FromMontgomery(test_set.scalars[i].ToMontgomery());
```
Comment (ashjeong) [inline tachyon/math/elliptic_curves/msm/algorithms/cuzk/cuzk_unittest.cc:81]: Can this be added to the first commit?

File: tachyon/math/finite_fields/prime_field_base.h [hunk 7]
```diff
@@ -165,11 +173,12 @@ class Copyable<
...
 
   static bool WriteTo(const T& prime_field, Buffer* buffer) {
-    if (s_is_in_montgomery) {
-      return buffer->Write(prime_field.ToMontgomery());
-    } else {
-      return buffer->Write(prime_field.ToBigInt());
+    if constexpr (T::Config::kUseMontgomery) {
+      if (s_is_in_montgomery) {
+        return buffer->Write(prime_field.value());
+      }
```
Comment (ashjeong) [inline tachyon/math/finite_fields/prime_field_base.h:179]:
```suggestion
    if constexpr (T::Config::kUseMontgomery && s_is_in_montgomery) {
      return buffer->Write(prime_field.value());
  ```
  Also change for the other 3 similar occurrences in this file

File: tachyon/math/finite_fields/prime_field_fallback.h [hunk 8]
```diff
@@ -96,7 +95,7 @@ class PrimeField<_Config, std::enable_if_t<!_Config::kIsSpecialPrime &&
     return PrimeField(big_int);
   }
 
-  constexpr static PrimeField FromMontgomery(const MontgomeryTy& mont) {
+  constexpr static PrimeField FromMontgomery(const BigInt<N>& mont) {
```
Comment (ashjeong) [inline tachyon/math/finite_fields/prime_field_fallback.h:98]: Why not use `BigIntTy`?

File: tachyon/math/finite_fields/prime_field_fallback.h [hunk 9]
```diff
@@ -38,7 +38,6 @@ class PrimeField<_Config, std::enable_if_t<!_Config::kIsSpecialPrime &&
 
   using Config = _Config;
   using BigIntTy = BigInt<N>;
-  using MontgomeryTy = BigInt<N>;
   using value_type = BigInt<N>;
```
Comment (ashjeong) [inline tachyon/math/finite_fields/prime_field_fallback.h:41]: The overlap in the type is quite confusing to me. What's the reason behind this?

File: tachyon/math/finite_fields/small_prime_field.h [hunk 10]
```diff
@@ -24,8 +24,7 @@ namespace tachyon::math {
 
 // A prime field is finite field GF(p) where p is a prime number.
 template <typename _Config>
-class PrimeField<_Config, std::enable_if_t<!_Config::kIsSpecialPrime &&
-                                           (_Config::kModulusBits <= 32) &&
+class PrimeField<_Config, std::enable_if_t<(_Config::kModulusBits <= 32) &&
```
Comment (ashjeong) [inline tachyon/math/finite_fields/small_prime_field.h:27]: Was the previous `!_Config::kIsSpecialPrime` unneeded here?

**Review by ashjeong — COMMENTED** [review]
Please change the following commits' titles/messages :
https://github.com/kroma-network/tachyon/pull/422/commits/180b74997fc9118b2fefd7994c945c6ddbe1d9c2
fix(math): instantiate `XXXGpu` with **the** same `PrimeField` class for sm…
https://github.com/kroma-network/tachyon/commit/c5d1e6f1a715ae4901f987638fd4d1dc7fa7c4aa
refac(math): separate `cpu.h.tpl` **into** `(small_)prime_field_cpu.h.tpl`
https://github.com/kroma-network/tachyon/pull/422/commits/a767713e9c389befc75389c1a8a232db6781ba8e
build: enable avx512 extension when **the flag is enabled**
https://github.com/kroma-network/tachyon/pull/422/commits/0188f04e9f7e7410dc3be7140e1c01d9c35cd627
**perf**(crypto): add round key to the first element only for the partial…
https://github.com/kroma-network/tachyon/pull/422/commits/469be685cc1abee1872366e5d97b377ca0103e5f
refac(math): separate `config.h.tpl` **into** `(small_)prime_field_config.h…
….tpl`...
Additionally, `BigInt<N>` value types have been changed to `uint32_t` for small prime fields
https://github.com/kroma-network/tachyon/pull/422/commits/31e2b08d7403f6b4b6a96a881ecda27b8f83c51c
Please change “As the range has been extended,...” to “As the prime field scope has increased,…”

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

**Review by dongchangYoo — APPROVED** [review]
Else, LGTM

File: tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl [hunk 11]
```diff
@@ -297,12 +306,17 @@ def _do_generate_prime_fields(
...
+    else:
+        prime_field_gpu_dep = "//tachyon/math/finite_fields:prime_field_gpu"
+
     tachyon_cc_library(
         name = "{}_gpu".format(name),
         hdrs = [":{}_gen_gpu_hdr".format(name)],
         deps = [
             ":{}_config".format(name),
-            "//tachyon/math/finite_fields:prime_field_gpu",
+            prime_field_gpu_dep,
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl:334]: When binary field is introduced, the code should be change as follows: 

https://github.com/kroma-network/tachyon/blob/4c639761dbe101cab3378688806cafa0be0982d3/tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl#L374-L389

```bazel
    if is_binary_prime_field:
        prime_field_gpu_dep = binary_prime_field_dep
    elif is_small_prime_field:
        prime_field_gpu_dep = small_prime_field_dep
    else:
        prime_field_gpu_dep = "//tachyon/math/finite_fields:prime_field_gpu"

    tachyon_cc_library(
        name = "{}_gpu".format(name),
        hdrs = [":{}_gen_gpu_hdr".format(name)],
        deps = [
            ":{}_config".format(name),
            prime_field_gpu_dep,
        ],
        **kwargs
    )
```

How do you think then?

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc [hunk 12]
```diff
@@ -340,16 +341,33 @@ int GenerationConfig::GenerateCpuHdr() const {
...
-                       {"%{config_header_path}",
-                        math::ConvertToConfigHdr(GetHdrPath()).value()},
-                       {"%{namespace}", ns_name},
-                       {"%{class}", class_name},
-                   });
+  if (is_small_field) {
+    CHECK(base::ReadFileToString(small_gpu_hdr_tpl_path, &tpl_content));
+  } else {
+    CHECK(base::ReadFileToString(gpu_hdr_tpl_path, &tpl_content));
+  }
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc:370]: When binary field is introduced, the code should be changed as follows:

https://github.com/kroma-network/tachyon/blob/4c639761dbe101cab3378688806cafa0be0982d3/tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc#L413-L420

```c++
  std::string tpl_content;
  if (is_binary_field) {
    CHECK(base::ReadFileToString(binary_gpu_hdr_tpl_path, &tpl_content));
  } else if (is_small_field) {
    CHECK(base::ReadFileToString(small_gpu_hdr_tpl_path, &tpl_content));
  } else {
    CHECK(base::ReadFileToString(gpu_hdr_tpl_path, &tpl_content));
  }
```

How do you think then?

File: tachyon/math/finite_fields/prime_field_gpu.h [hunk 13]
```diff
@@ -109,8 +109,8 @@ class PrimeFieldGpu final : public PrimeFieldBase<PrimeFieldGpu<_Config>> {
 
   static void Init() { VLOG(1) << Config::kName << " initialized"; }
 
-  __host__ __device__ const value_type& value() const { return value_; }
-  __host__ __device__ size_t GetLimbSize() const { return N; }
+  constexpr const value_type& value() const { return value_; }
+  constexpr size_t GetLimbSize() const { return N; }
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/prime_field_gpu.h:112]: Can still work on the device!

File: tachyon/math/finite_fields/small_prime_field.h (hunk 10 above)
Comment (chokobole) [reply tachyon/math/finite_fields/small_prime_field.h:27]: Yes, I think the flag `kIsSpecialPrime` is changed to `kUseAsm` in this PR. Since there's no assembly code for a Small Prime Field, it doesn't need to have this condition any more.

File: tachyon/math/finite_fields/prime_field_fallback.h (hunk 8 above)
Comment (chokobole) [reply tachyon/math/finite_fields/prime_field_fallback.h:98]: `BigIntTy` is the type to be used to use the type outside of this PrimeField. Although using `BigIntTy` also makes sense, the rest of the code here uses `BigInt<N>`.

File: tachyon/math/finite_fields/prime_field_fallback.h (hunk 9 above)
Comment (chokobole) [reply tachyon/math/finite_fields/prime_field_fallback.h:41]: Do you mean why there exist the type `MontgomeryTy`? If so, I make this type to explicitly denote the return type for `ToMontgomery()`.

File: tachyon/math/finite_fields/prime_field_base.h (hunk 7 above)
Comment (chokobole) [reply tachyon/math/finite_fields/prime_field_base.h:179]: We can't do this. since `s_is_in_montgomery` is not determined at compile time.

File: tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc [hunk 14]
```diff
@@ -303,8 +304,16 @@ int GenerationConfig::GenerateConfigHdr() const {
...
+  size_t num_bits = GetNumBits(m);
+  bool is_small_field = num_bits <= 32;
+
   std::string tpl_content;
-  CHECK(base::ReadFileToString(cpu_hdr_tpl_path, &tpl_content));
+  if (is_small_field) {
+    CHECK(base::ReadFileToString(small_cpu_hdr_tpl_path, &tpl_content));
+  } else {
+    CHECK(base::ReadFileToString(cpu_hdr_tpl_path, &tpl_content));
+  }
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc:330]: When binary field is introduced, the code should be changed as follows:

https://github.com/kroma-network/tachyon/blob/4c639761dbe101cab3378688806cafa0be0982d3/tachyon/math/finite_fields/generator/prime_field_generator/prime_field_generator.cc#L370-L377

How do you think then?

File: tachyon/math/finite_fields/goldilocks/goldilocks_prime_field_x86_special.h
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/goldilocks/goldilocks_prime_field_x86_special.h:1]: `BigIntTy` is to be used outside of this file!

File: tachyon/math/elliptic_curves/msm/algorithms/cuzk/cuzk_unittest.cc (hunk 6 above)
Comment (chokobole) [reply tachyon/math/elliptic_curves/msm/algorithms/cuzk/cuzk_unittest.cc:81]: I think it's possible. But isn't it more likely removing montgomery conversions?

File: tachyon/math/finite_fields/prime_field_fallback.h (hunk 9 above)
Comment (ashjeong) [reply tachyon/math/finite_fields/prime_field_fallback.h:41]: I was confused on why there were three separate aliases for the same type. Are the values held inside different regardless of type?

File: tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl (hunk 11 above)
[the remark this answers is not in the document]
Comment (ashjeong) [reply tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl:334]: Though I'm not sure on what the difference is when binary field is introduced, I still think a ternary operator may be cleaner, especially since  some if/else statements using `is_small_prime_field` are stated before, so it feels like this if/else is possibly redundant as well.

File: tachyon/math/elliptic_curves/msm/algorithms/cuzk/cuzk_unittest.cc (hunk 6 above)
Comment (ashjeong) [reply tachyon/math/elliptic_curves/msm/algorithms/cuzk/cuzk_unittest.cc:81]: That's fair. I suppose this isn't an explicit cast but more of a conversion.

File: tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl (hunk 11 above)
[the remark this answers is not in the document]
Comment (ashjeong) [reply tachyon/math/finite_fields/generator/prime_field_generator/build_defs.bzl:334]: In the case that the if statements include more than 2 branches, a ternary operator should not be used, so as you said, it makes sense that this should be a series of if/else statements. I'll resolve this.

File: tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h [hunk 15]
```diff
@@ -54,19 +47,19 @@ struct PoseidonSpongeBase : public FieldBasedCryptographicSponge<Derived> {
 
     size_t full_rounds_over_2 = config.full_rounds / 2;
     for (size_t i = 0; i < full_rounds_over_2; ++i) {
-      ApplyARK(i);
+      derived.ApplyARK(i, true);
```
Comment (Insun35) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h:50]:
```suggestion
      derived.ApplyARK(i, /*is_full_round=*/true);
```

File: tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h [hunk 16]
```diff
@@ -54,19 +47,19 @@ struct PoseidonSpongeBase : public FieldBasedCryptographicSponge<Derived> {
...
       ApplySBox(true);
       derived.ApplyMix(true);
     }
     for (size_t i = full_rounds_over_2;
          i < full_rounds_over_2 + config.partial_rounds; ++i) {
-      ApplyARK(i);
+      derived.ApplyARK(i, false);
```
Comment (Insun35) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h:56]:
```suggestion
      derived.ApplyARK(i, /*is_full_round=*/false);
```

File: tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h [hunk 17]
```diff
@@ -54,19 +47,19 @@ struct PoseidonSpongeBase : public FieldBasedCryptographicSponge<Derived> {
...
       ApplySBox(false);
       derived.ApplyMix(false);
     }
     for (size_t i = full_rounds_over_2 + config.partial_rounds;
          i < config.partial_rounds + config.full_rounds; ++i) {
-      ApplyARK(i);
+      derived.ApplyARK(i, true);
```
Comment (Insun35) [inline tachyon/crypto/hashes/sponge/poseidon/poseidon_sponge_base.h:62]:
```suggestion
      derived.ApplyARK(i, /*is_full_round=*/true);
```

File: tachyon/c/math/elliptic_curves/msm/msm_gpu_unittest.cc [hunk 18]
```diff
@@ -44,8 +44,10 @@ TEST_P(MSMGpuTest, MSMPoint2) {
   for (const VariableBaseMSMTestSet<bn254::G1AffinePoint>& t :
        this->test_sets_) {
     std::unique_ptr<tachyon_bn254_g1_jacobian> ret;
-    std::vector<Point2<BigInt<4>>> bases = base::CreateVector(
-        t.bases.size(), [&t](size_t i) { return t.bases[i].ToMontgomery(); });
+    std::vector<Point2<bn254::Fq>> bases =
+        base::CreateVector(t.bases.size(), [&t](size_t i) {
+          return Point2<bn254::Fq>(t.bases[i].x(), t.bases[i].y());
```
Comment (Insun35) [inline tachyon/c/math/elliptic_curves/msm/msm_gpu_unittest.cc:49]: Don't we need to handle the infinity case where x and y are all zero?

Comment (chokobole) [reply tachyon/c/math/elliptic_curves/msm/msm_gpu_unittest.cc:49]: We don't need checking here!

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM
